compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/HeaderProcessor.kt

```diff
@@ -37,39 +39,123 @@ internal data class LoopReplacement(
+    init {
+        with(builder) {
+            // For this loop:
+            //
+            //   for (i in first()..last() step step())
+            //
+            // We need to cast first(), last(). and step() to conform to the progression type so
+            // that operations on the induction variable within the loop are more efficient.
+            //
+            // In the above example, if first() is a Long and last() is an Int, this creates a
+            // LongProgression so last() should be cast to a Long.
+            inductionVariable = scope.createTemporaryVariable(
+                headerInfo.first.castIfNecessary(
+                    headerInfo.progressionType.elementType(context.irBuiltIns),
+                    headerInfo.progressionType.elementCastFunctionName
+                ),
+                nameHint = "inductionVariable",
+                isMutable = true
+            )
 
-    /** Builds a new loop from the old loop. */
-    abstract fun buildLoop(builder: DeclarationIrBuilder, oldLoop: IrLoop, newBody: IrExpression?): LoopReplacement
+            // Due to features of PSI2IR we can obtain nullable arguments here while actually
+            // they are non-nullable (the frontend takes care about this). So we need to cast
+            // them to non-nullable.
+            // TODO: Confirm if casting to non-nullable is still necessary
+            val last = ensureNotNullable(
+                headerInfo.last.castIfNecessary(
+                    headerInfo.progressionType.elementType(context.irBuiltIns),
+                    headerInfo.progressionType.elementCastFunctionName
+                )
+            )
+
+            lastVariableIfCanCacheLast = if (headerInfo.canCacheLast) {
+                scope.createTemporaryVariable(
+                    last,
+                    nameHint = "last"
+                )
+            } else null
+
+            lastExpression = if (headerInfo.canCacheLast) irGet(lastVariableIfCanCacheLast!!) else last
+
+            stepVariable = headerInfo.progressionType.stepType(context.irBuiltIns).let {
+                scope.createTemporaryVariable(
+                    ensureNotNullable(
+                        headerInfo.step.castIfNecessary(
+                            it,
+                            headerInfo.progressionType.stepCastFunctionName
+                        )
+                    ),
+                    nameHint = "step",
+                    irType = it
+                )
+            }
+        }
+    }
+
+    // This cannot be declared and initialized directly in the constructor. At the time of the base class (ForLoopHeader) constructor
+    // execution, the `headerInfo` property overridden in the derived class (e.g., NumericForLoopHeader) is not yet initialized.
+    // Therefore, we must use the `headerInfo` constructor parameter in the "init" block above instead of using the property.
+    protected open val headerInfo: NumericHeaderInfo = headerInfo
```

Comment:
Would class generalization solve this problem (e.g. add generic parameters)?

Reply:
@max-kammerer That works! Fixed.